File: 1283. Find the Smallest Divisor Given a Threshold/Solution.java
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int right = Integer.MIN_VALUE;

        for(int i=0;i<nums.length;i++){
            right = Math.max(right,nums[i]);
        }

        int ans = right;

        int left = 1;

        while(left<=right){
            int mid = left + (right-left)/2;

            if(isPossible(nums,mid,threshold)){
                ans = mid;
                right = mid-1;
            }
            else{
                left = mid+1;
            }
        }
        return ans;
    }
    private boolean isPossible(int[] nums,int d,int t){
        int total = 0;
        for(int i=0;i<nums.length;i++){
            int ceil = (nums[i]+d-1)/d;
            total += ceil;
        }

        return total<=t;
    }
}

File: 1283. Find the Smallest Divisor Given a Threshold/Solution.py
class Solution(object):
    def smallestDivisor(self, nums, threshold):
        """
        :type nums: List[int]
        :type threshold: int
        :rtype: int
        """
        right = max(nums)
        left = 1
        ans = right

        while left<=right:
            mid = left + (right-left)//2

            if self.isPossible(nums,mid,threshold):
                ans = mid
                right = mid-1
            else:
                left = mid+1
        return ans

    def isPossible(self,nums,d,t):
        total = 0

        for i in nums:
            total += ((i+d-1)/d)

        return total<=t

        